### ISSUE TITLE
7$: Crappy error message for (g()) => {}

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#322]
quick-lint-js gives an unhelpful diagnostic for the following code:
```js
let f = (g()) => {};
```
```
hello.js:1:10: error: missing operator between expression and arrow function [E063]
```

**Comment by strager** [comment]
```js
let f = (g()) => {};  // should say invalid parameter?
let f = (g()) => 42;  // should say => is unexpected (should maybe be <= instead)
```

**Comment by singalhimanshu** [comment]
> ```js
> let f = (g()) => {};  // should say invalid parameter?
> let f = (g()) => 42;  // should say => is unexpected (should maybe be <= instead)
> ```

Shouldn't we produce the same error in both the cases? Some of the js engines reports the following errors:
```shell
#### JavaScriptCore

SyntaxError: Unexpected token '=>'

#### SpiderMonkey

SyntaxError: missing ) after formal parameters:

#### V8

SyntaxError: Invalid destructuring assignment target
```

And currently `quick-lint-js` gives the following error:
```shell
main.js:2:15: error: unexpected '=>' [E160]
main.js:2:10: note: expected parameter for arrow function, but got an expression instead [E160]
```

I think the parsing for second case is correct we should just change it for the first case (`let f = (g()) => {};`)?

**Comment by strager** [comment]
> I think the parsing for second case is correct

Right. quick-lint-js should and does report E160 for `let f = (g()) => 42;`.

> we should just change it for the first case (`let f = (g()) => {};`)?

Correct.

> Shouldn't we produce the same error in both the cases?

No. Here are examples motivating quick-lint-js' current behavior:

```js
// programmer forgot ',' between 'path' and '(err, data)':
fs.readFile(path (err, data) => { console.log(data); })

// programmer wrote '=>' instead of '>=':
if (getMapSize() => 3) { console.error("too many items"); }
// If the '=>' was followed by a '{', then '>=' likely would make no sense.
```

(We should include these examples in docs/errors/...)

**Comment by strager** [comment]
> (We should include these examples in docs/errors/...)

Done: https://github.com/quick-lint/quick-lint-js/commit/d782731820b78a5598da0e66f9ce5a561290f2d4
